src/main/java/net/neoforged/neoforge/common/PointOfInterestUnloading.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common;
+
+import net.minecraft.core.SectionPos;
+import net.minecraft.world.entity.ai.village.poi.PoiManager;
+import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.chunk.ChunkAccess;
+
+public class PointOfInterestUnloading {
+    /**
+     * The goal here is to fix the POI memory leak that happens due to
+     * {@link net.minecraft.world.level.chunk.storage.SectionStorage#storage} field never
+     * actually removing POIs long after they become irrelevant. We do it here in chunk unload event
+     * so that chunk that are fully unloaded now gets the POI removed from the POI cached storage map.
+     */
+    public static void onChunkUnload(PoiManager poiManager, ChunkAccess chunkAccess) {
```

Comment:
This is small enough that I'd just put it in `CommonHooks` and change the name to `handlePOIChunkUnload`.